geckomat/change_model/applyKcatConstraints.m

```diff
@@ -80,90 +41,60 @@
-            case 'min'
-                nonZeroKcat = updateKcats(sameSubunit);
-                updateKcats(sameSubunit(1))=min(nonZeroKcat(nonZeroKcat>0),[],'omitnan');
-            case 'median'
-                nonZeroKcat = updateKcats(sameSubunit);                
-                updateKcats(sameSubunit(1))=median(nonZeroKcat(nonZeroKcat>0),'omitnan');
-        end
-        %Set new kcat value at the position of the first occurence of the
-        %subunit, and remove all other kcats for this subunit (for this rxn)
-        sameSubunit(1)=[];
-        updateKcats(sameSubunit)=0; %Remove non-summarized kcat value
+if ~model.ec.geckoLight %For normal GECKO formulation, where each enzyme is explicitly considered
+    %Make vector where column 1 = rxn; 2 = enzyme; 3 = subunit copies; 4 = kcat
+    newKcats=zeros(numel(updateRxns)*10,4);
+    updateRxns=find(updateRxns);
+    kcatFirst=0;
+    for i=1:numel(updateRxns)
+        enzymes   = find(model.ec.rxnEnzMat(i,:));
+        kcatLast  = kcatFirst+numel(enzymes);
+        kcatFirst = kcatFirst+1;
+        newKcats(kcatFirst:kcatLast,1) = updateRxns(i);
+        newKcats(kcatFirst:kcatLast,2) = enzymes;
+        newKcats(kcatFirst:kcatLast,3) = model.ec.rxnEnzMat(i,enzymes);
+        newKcats(kcatFirst:kcatLast,4) = model.ec.kcat(i);
+        kcatFirst = kcatLast;
     end
-    multipleKcatEntries=multipleKcatEntries(firstIdx);
-    switch kcatComplex
-        case 'max'
-            updateKcats(multipleKcatEntries) = max(updateKcats(multipleKcatEntries),[],'omitnan');
-        case 'min'
-            nonZeroKcat = updateKcats(multipleKcatEntries);
-            updateKcats(multipleKcatEntries) = min(nonZeroKcat(nonZeroKcat>0),[],'omitnan');
-        case 'median'
-            nonZeroKcat = updateKcats(multipleKcatEntries);
-            updateKcats(multipleKcatEntries) = median(nonZeroKcat(nonZeroKcat>0),'omitnan');
+    newKcats(kcatLast+1:end,:)=[];
+    
+    newKcats(:,4) = newKcats(:,4) * 3600; %per second -> per hour
+    newKcats(:,4) = newKcats(:,4).^-1; %Inverse: hours per reaction
+    newKcats(:,4) = newKcats(:,4)*1000; %In umol instead of mmol
+    newKcats(:,4) = newKcats(:,3).*newKcats(:,4); %Multicopy subunits.
+    %Unit of enzyme usage is umol/gDW/h, while metabolic flux is in mmol/gDW/h.
+    %This prevents very low fluxes.
+    
+    %Replace rxns and enzymes with their location in model
+    [~,newKcats(:,1)] = ismember(model.ec.rxns(newKcats(:,1)),model.rxns);
+    [~,newKcats(:,2)] = ismember(strcat('prot_',model.ec.enzymes(newKcats(:,2))),model.mets);
+    linearIndices     = sub2ind(size(model.S),newKcats(:,2),newKcats(:,1));
+    model.S(linearIndices) = -newKcats(:,4); %Substrate = negative
+    
+else %GECKO light formulation, where prot_pool represents all usages
+    newKcats=zeros(numel(updateRxns),1);
+    for i=1:numel(updateRxns)
+        %TODO:The current code is incorrect, it takes the sum of all
```

Comment:
So, we need to have all alternative kcats with us here (i.e., the kcat for each isozyme), and use the largest of those. So, the model.ec.kcat needs to support multiple numbers for Gecko light. These structures could be different for Gecko light, I don't think there is much common code between the two ways to lay out the model?

Reply:
Various functions can gather alternative kcats, before `selectKcatValue` selects the appropriate value (for DLKcat this is the max value for different substrates, for GeckoLight this is the min MW/kcat. These selected values are then entered into `model.ec.kcat`, so no need to support multiple kcat values there.

Reply:
Ok, sounds good